apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java

```diff
@@ -427,7 +427,12 @@ public void asRootSpan(Sampler sampler) {
         transactionId.copyFrom(id);
         if (sampler.isSampled(traceId)) {
             flags = FLAG_RECORDED;
-            traceState.set(sampler.getSampleRate(), sampler.getTraceStateHeader());
+
+            if (Double.isNaN(traceState.getSampleRate())) {
+                // sample rate could already have been set through headers
+                // we have to make sure to only use sampler rate as fallback only.
```

Comment:
How come it could have been set through headers? Isn't `asRootSpan()` only invoked when no trace context headers are received (i.e. root)?

Reply:
1. yes, first wins is a sensible strategy here.
2. logging with `debug` is fine as it's an exceptional case that would only be investigated.
3. yes, we should only propagate valid header and de-duplicate.

Even if we should not call `TraceState#set` twice, having a consistent behavior with `TraceState.addTextHeader()` would be nice, thus logging both with `debug` seems the right option.